File: software/openroam-core/openroam_core/roamk.py
"""
RoamK - OpenRoam Data Model

Similar to Signal K for marine vessels, RoamK provides a unified
data model for all vehicle systems.
"""

from dataclasses import dataclass, field
from typing import Optional
from datetime import datetime


@dataclass
class Location:
    """GPS location data."""

    latitude: float = 0.0
    longitude: float = 0.0
    altitude: Optional[float] = None
    speed: Optional[float] = None
    heading: Optional[float] = None
    satellites: Optional[int] = None


@dataclass
class BatteryBank:
    """Battery bank state."""

    voltage: float = 0.0
    current: float = 0.0
    soc: float = 0.0
    temp: Optional[float] = None
    capacity_ah: Optional[float] = None


@dataclass
class PowerState:
    """Power system state."""

    house_battery: BatteryBank = field(default_factory=BatteryBank)
    chassis_battery: BatteryBank = field(default_factory=BatteryBank)
    solar_voltage: float = 0.0
    solar_current: float = 0.0
    solar_watts: float = 0.0
    solar_daily_wh: float = 0.0
    solar_lifetime_kwh: float = 0.0
    shore_connected: bool = False
    shore_voltage: float = 0.0
    shore_amps: float = 0.0
    alternator_charging: bool = False
    alternator_amps: float = 0.0


@dataclass
class TankLevel:
    """Tank level data."""

    level: float = 0.0
    gallons: Optional[float] = None
    capacity: Optional[float] = None


@dataclass
class TanksState:
    """All tank levels."""

    fresh: TankLevel = field(default_factory=TankLevel)
    grey: TankLevel = field(default_factory=TankLevel)
    black: TankLevel = field(default_factory=TankLevel)
    propane: TankLevel = field(default_factory=TankLevel)
    fuel: TankLevel = field(default_factory=TankLevel)


@dataclass
class ClimateZone:
    """Climate zone data."""

    temperature: float = 0.0
    setpoint: Optional[float] = None
    humidity: Optional[float] = None


@dataclass
class ClimateState:
    """Climate system state."""

    interior_temp: float = 0.0
    interior_humidity: float = 0.0
    exterior_temp: float = 0.0
    exterior_humidity: float = 0.0
    zones: dict[str, ClimateZone] = field(default_factory=dict)
    hvac_mode: str = "off"  # heat, cool, auto, off
    hvac_running: bool = False
    hvac_fan_speed: int = 0


@dataclass
class EngineState:
    """Engine/vehicle data from OBD-II."""

    rpm: int = 0
    coolant_temp: int = 0
    oil_temp: int = 0
    oil_pressure: int = 0
    trans_temp: int = 0
    throttle: int = 0
    load: int = 0
    fuel_rate: float = 0.0
    mpg_instant: float = 0.0
    mpg_average: float = 0.0
    dtc_codes: list[str] = field(default_factory=list)
    check_engine: bool = False


@dataclass
class SafetyState:
    """Safety system state."""

    smoke_status: str = "ok"  # ok, warning, alarm
    co_ppm: int = 0
    co_status: str = "ok"
    propane_ppm: int = 0
    propane_status: str = "ok"
    doors: dict[str, str] = field(default_factory=dict)  # door: open/closed
    alarm_armed: bool = False
    alarm_triggered: bool = False


@dataclass
class VehicleState:
    """Vehicle state."""

    location: Location = field(default_factory=Location)
    speed: float = 0.0
    heading: float = 0.0
    odometer: float = 0.0
    leveling_pitch: float = 0.0
    leveling_roll: float = 0.0
    leveling_state: str = "travel"  # leveling, level, travel


@dataclass
class MaintenanceItem:
    """Maintenance reminder."""

    service: str = ""
    due_miles: Optional[int] = None
    due_date: Optional[str] = None


@dataclass
class MaintenanceState:
    """Maintenance tracking."""

    oil_life: int = 100
    next_oil_miles: int = 5000
    tire_rotation_due: bool = False
    upcoming: list[MaintenanceItem] = field(default_factory=list)


@dataclass
class RoamKState:
    """Complete RoamK state."""

    vehicle: VehicleState = field(default_factory=VehicleState)
    power: PowerState = field(default_factory=PowerState)
    tanks: TanksState = field(default_factory=TanksState)
    climate: ClimateState = field(default_factory=ClimateState)
    engine: EngineState = field(default_factory=EngineState)
    safety: SafetyState = field(default_factory=SafetyState)
    maintenance: MaintenanceState = field(default_factory=MaintenanceState)
    last_update: datetime = field(default_factory=datetime.now)


class RoamK:
    """RoamK data store."""

    def __init__(self) -> None:
        self.state = RoamKState()
        self._callbacks: list[callable] = []

    def update_path(self, path: str, value: any) -> None:
        """Update a value at the given path."""
        parts = path.split(".")
        obj = self.state

        for part in parts[:-1]:
            if hasattr(obj, part):
                obj = getattr(obj, part)
            else:
                return

        if hasattr(obj, parts[-1]):
            setattr(obj, parts[-1], value)
            self.state.last_update = datetime.now()
            self._notify()

    def get_path(self, path: str) -> any:
        """Get a value at the given path."""
        parts = path.split(".")
        obj = self.state

        for part in parts:
            if hasattr(obj, part):
                obj = getattr(obj, part)
            else:
                return None

        return obj

    def subscribe(self, callback: callable) -> None:
        """Subscribe to state changes."""
        self._callbacks.append(callback)

    def unsubscribe(self, callback: callable) -> None:
        """Unsubscribe from state changes."""
        self._callbacks.remove(callback)

    def _notify(self) -> None:
        """Notify all subscribers of state change."""
        for callback in self._callbacks:
            callback(self.state)

    def to_dict(self) -> dict:
        """Convert state to dictionary."""
        from dataclasses import asdict

        return asdict(self.state)

File: software/openroam-dashboard/src/lib/stores/roamk.ts
/**
 * RoamK - OpenRoam Data Store
 * Similar to Signal K for marine vessels, RoamK provides a unified
 * data model for all vehicle systems.
 */

import { writable, derived, type Readable } from 'svelte/store';

// Type definitions for RoamK data model
export interface Location {
	latitude: number;
	longitude: number;
	altitude?: number;
	speed?: number;
	heading?: number;
	satellites?: number;
}

export interface BatteryBank {
	voltage: number;
	current: number;
	soc: number;
	temp?: number;
	capacity_ah?: number;
}

export interface PowerState {
	batteries: {
		house?: BatteryBank;
		chassis?: BatteryBank;
		aux?: BatteryBank;
	};
	solar: {
		voltage: number;
		current: number;
		watts: number;
		daily_wh: number;
		lifetime_kwh?: number;
	};
	shore: {
		connected: boolean;
		voltage: number;
		amps: number;
	};
	alternator: {
		charging: boolean;
		amps: number;
	};
}

export interface TankLevel {
	level: number;
	gallons?: number;
	capacity?: number;
}

export interface TanksState {
	fresh: TankLevel;
	grey: TankLevel;
	black: TankLevel;
	propane: TankLevel & { pounds?: number };
	fuel: TankLevel & { range_miles?: number };
}

export interface ClimateZone {
	temperature: number;
	setpoint?: number;
	humidity?: number;
}

export interface ClimateState {
	interior: { temperature: number; humidity: number };
	exterior: { temperature: number; humidity: number };
	zones: Record<string, ClimateZone>;
	hvac: {
		mode: 'heat' | 'cool' | 'auto' | 'off';
		running: boolean;
		fan_speed: number;
	};
}

export interface EngineState {
	rpm: number;
	coolant_temp: number;
	oil_temp: number;
	oil_pressure: number;
	trans_temp: number;
	throttle: number;
	load: number;
	fuel_rate: number;
	mpg_instant: number;
	mpg_average: number;
	dtc_codes: string[];
	check_engine: boolean;
}

export interface SafetyState {
	smoke: { status: 'ok' | 'warning' | 'alarm' };
	co: { ppm: number; status: 'ok' | 'warning' | 'alarm' };
	propane: { ppm: number; status: 'ok' | 'warning' | 'alarm' };
	doors: Record<string, 'open' | 'closed'>;
	alarm: { armed: boolean; triggered: boolean };
}

export interface VehicleState {
	location: Location;
	speed: number;
	heading: number;
	odometer: number;
	leveling?: {
		pitch: number;
		roll: number;
		state: 'leveling' | 'level' | 'travel';
	};
}

export interface MaintenanceItem {
	service: string;
	due_miles?: number;
	due_date?: string;
}

export interface MaintenanceState {
	oil_life: number;
	next_oil_miles: number;
	tire_rotation_due: boolean;
	upcoming: MaintenanceItem[];
}

export interface SystemNode {
	name: string;
	online: boolean;
	cpu: number;
	memory: number;
	temp: number;
	disk: number;
}

export interface NetworkState {
	wifi_clients: number;
	cellular_signal: number;
	carrier: string;
	data_used_gb: number;
	starlink?: {
		connected: boolean;
		speed_mbps: number;
		obstructions: number;
	};
}

// Complete RoamK state
export interface RoamKState {
	vehicle: VehicleState;
	power: PowerState;
	tanks: TanksState;
	climate: ClimateState;
	engine: EngineState;
	safety: SafetyState;
	maintenance: MaintenanceState;
	network: NetworkState;
	system: {
		nodes: SystemNode[];
	};
	lastUpdate: number;
}

// Default initial state - populated by MQTT data from hardware
const initialState: RoamKState = {
	vehicle: {
		location: { latitude: 0, longitude: 0 },
		speed: 0,
		heading: 0,
		odometer: 0
	},
	power: {
		batteries: {
			house: { voltage: 0, current: 0, soc: 0 },
			chassis: { voltage: 0, current: 0, soc: 0 }
		},
		solar: { voltage: 0, current: 0, watts: 0, daily_wh: 0 },
		shore: { connected: false, voltage: 0, amps: 0 },
		alternator: { charging: false, amps: 0 }
	},
	tanks: {
		fresh: { level: 0 },
		grey: { level: 0 },
		black: { level: 0 },
		propane: { level: 0 },
		fuel: { level: 0 }
	},
	climate: {
		interior: { temperature: 0, humidity: 0 },
		exterior: { temperature: 0, humidity: 0 },
		zones: {},
		hvac: { mode: 'off', running: false, fan_speed: 0 }
	},
	engine: {
		rpm: 0,
		coolant_temp: 0,
		oil_temp: 0,
		oil_pressure: 0,
		trans_temp: 0,
		throttle: 0,
		load: 0,
		fuel_rate: 0,
		mpg_instant: 0,
		mpg_average: 0,
		dtc_codes: [],
		check_engine: false
	},
	safety: {
		smoke: { status: 'ok' },
		co: { ppm: 0, status: 'ok' },
		propane: { ppm: 0, status: 'ok' },
		doors: {},
		alarm: { armed: false, triggered: false }
	},
	maintenance: {
		oil_life: 100,
		next_oil_miles: 5000,
		tire_rotation_due: false,
		upcoming: []
	},
	network: {
		wifi_clients: 0,
		cellular_signal: 0,
		carrier: '',
		data_used_gb: 0
	},
	system: {
		nodes: []
	},
	lastUpdate: Date.now()
};

// Create the main store
function createRoamKStore() {
	const { subscribe, set, update } = writable<RoamKState>(initialState);

	return {
		subscribe,
		set,
		update,

		// Update a specific path in the state
		updatePath: (path: string, value: unknown) => {
			update((state) => {
				const parts = path.split('.');
				let current: Record<string, unknown> = state as unknown as Record<string, unknown>;

				for (let i = 0; i < parts.length - 1; i++) {
					if (!(parts[i] in current)) {
						current[parts[i]] = {};
					}
					current = current[parts[i]] as Record<string, unknown>;
				}

				current[parts[parts.length - 1]] = value;
				state.lastUpdate = Date.now();
				return state;
			});
		},

		// Reset to initial state
		reset: () => set(initialState)
	};
}

// Export the main store
export const roamk = createRoamKStore();

// Derived stores for specific sections
export const vehicle: Readable<VehicleState> = derived(roamk, ($roamk) => $roamk.vehicle);
export const power: Readable<PowerState> = derived(roamk, ($roamk) => $roamk.power);
export const tanks: Readable<TanksState> = derived(roamk, ($roamk) => $roamk.tanks);
export const climate: Readable<ClimateState> = derived(roamk, ($roamk) => $roamk.climate);
export const engine: Readable<EngineState> = derived(roamk, ($roamk) => $roamk.engine);
export const safety: Readable<SafetyState> = derived(roamk, ($roamk) => $roamk.safety);
export const maintenance: Readable<MaintenanceState> = derived(roamk, ($roamk) => $roamk.maintenance);
export const network: Readable<NetworkState> = derived(roamk, ($roamk) => $roamk.network);

// Connection status
export const connected = writable(false);
